### PR TITLE
Middleware restructure

### PR DESCRIPTION
**Pull request by adam-fowler**
Changes required for middleware changes. See https://github.com/soto-project/soto/pull/687 for more details

### CODE REVIEW

File: Sources/SotoCodeGeneratorLib/Templates/api.swift [hunk 1]
```diff
@@ -52,14 +52,22 @@ extension Templates {
     {{/regionalized}}
         ///     - partition: AWS partition where service resides, standard (.aws), china (.awscn), government (.awsusgov).
         ///     - endpoint: Custom endpoint URL to use instead of standard AWS servers
+    {{^middlewareClass}}
+        ///     - middleware: Middleware chain used to edit requests before they sent and responses before they are decoded
```
Comment (0xTim) [inline Sources/SotoCodeGeneratorLib/Templates/api.swift:56]:
```suggestion
        ///     - middleware: Middleware chain used to edit requests before they are sent and responses before they are decoded 
```

File: Sources/SotoCodeGeneratorLib/Templates/api.swift [hunk 2]
```diff
@@ -127,6 +119,111 @@ extension Templates {
...
+    {{#middlewareClass}}
+        /// Initialize the {{name}} client
+        /// - parameters:
+        ///     - client: AWSClient used to process requests
+    {{#regionalized}}
+        ///     - region: Region of server you want to communicate with. This will override the partition parameter.
+    {{/regionalized}}
+        ///     - partition: AWS partition where service resides, standard (.aws), china (.awscn), government (.awsusgov).
+        ///     - endpoint: Custom endpoint URL to use instead of standard AWS servers
+        ///     - middleware: Middleware chain used to edit requests before they sent and responses before they are decoded
```
Comment (0xTim) [inline Sources/SotoCodeGeneratorLib/Templates/api.swift:131]:
```suggestion
        ///     - middleware: Middleware chain used to edit requests before they are sent and responses before they are decoded 
```

File: Sources/SotoCodeGeneratorLib/Templates/api.swift [hunk 3]
```diff
@@ -165,7 +263,7 @@ extension Templates {
     }
 
     extension {{ name }} {
-        /// Initializer required by `AWSService.with(middlewares:timeout:byteBufferAllocator:options)`. You are not able to use this initializer directly as there are no {{scope}}
+        /// Initializer required by `AWSService.with(middlewares:timeout:byteBufferAllocator:options)`. You are not able to use this initializer directly as there are no public
```
Comment (0xTim) [inline Sources/SotoCodeGeneratorLib/Templates/api.swift:266]: Not?
```suggestion
        /// Initializer required by `AWSService.with(middlewares:timeout:byteBufferAllocator:options)`. You are not able to use this initializer directly as there are not public
```

**Review by 0xTim — CHANGES_REQUESTED** [review]
Some typos

**Review by 0xTim — APPROVED** [review]

File: Sources/SotoCodeGeneratorLib/Templates/api.swift [hunk 4]
```diff
@@ -165,7 +263,7 @@ extension Templates {
     }
 
     extension {{ name }} {
-        /// Initializer required by `AWSService.with(middlewares:timeout:byteBufferAllocator:options)`. You are not able to use this initializer directly as there are no {{scope}}
+        /// Initializer required by `AWSService.with(middlewares:timeout:byteBufferAllocator:options)`. You are not able to use this initializer directly as there are not public
```
Comment (0xTim) [inline Sources/SotoCodeGeneratorLib/Templates/api.swift:266]:
```suggestion
        /// Initializer required by `AWSService.with(middlewares:timeout:byteBufferAllocator:options)`. You are not able to use this initializer directly as these are not public
```

**Review by 0xTim — COMMENTED** [review]
Should it be this?

File: Sources/SotoCodeGeneratorLib/Templates/api.swift (hunk 4 above)
Comment (adam-fowler) [reply Sources/SotoCodeGeneratorLib/Templates/api.swift:266]: Read the next line, `there are not public intializers for AWSServiceConfig.Patch`

Comment (0xTim) [reply Sources/SotoCodeGeneratorLib/Templates/api.swift:266]: Ah the mobile app

**Pull request merged**